projects/novo-elements/src/utils/app-bridge/AppBridge.ts

```diff
@@ -151,129 +164,160 @@ export class AppBridge {
+        const index = this._registeredFrames.findIndex((frame) => frame.data.id === event.data.id);
+        if (index !== -1) {
+          this._registeredFrames.splice(index, 1);
+        }
+        return this.close(event.data).then(success => ({ success }));
+      },
+      // Refresh
+      [MESSAGE_TYPES.REFRESH]: (event: PostRobotEvent<any>) => {
+        return this.refresh(event.data).then((success) => {
+          return { success };
+        });
+      },
+      // PIN
+      [MESSAGE_TYPES.PIN]: (event: PostRobotEvent<any>) => {
+        return this.pin(event.data).then((success) => {
+          return { success };
+        });
+      },
+      // PING
+      [MESSAGE_TYPES.PING]: (event: PostRobotEvent<any>) => {
+        return this.httpGET('ping', undefined, event.data.origin).then((result) => {
+          return { data: result.data, error: result.error };
+        });
+      },
+      // REQUEST_DATA
+      [MESSAGE_TYPES.REQUEST_DATA]: (event: PostRobotEvent<any>) => {
+        return this.requestData(event.data).then((result) => {
+          return { data: result.data, error: result.error };
+        });
+      },
+      // CALLBACKS
+      [MESSAGE_TYPES.CALLBACK]: (event: PostRobotEvent<any>) => {
+        return this.callback(event.data).then((success) => {
+          return { success };
+        });
+      },
+      // HTTP-GET
+      [MESSAGE_TYPES.HTTP_GET]: (event: PostRobotEvent<any>) => {
+        return this.httpGET(event.data.relativeURL, undefined, event.data.origin).then((result) => {
+          return { data: result.data, error: result.error };
+        });
+      },
+      // HTTP-POST
+      [MESSAGE_TYPES.HTTP_POST]: (event: PostRobotEvent<any>) => {
+        return this.httpPOST(event.data.relativeURL, event.data.data, undefined, event.data.origin).then((result) => {
+          return { data: result.data, error: result.error };
+        });
+      },
+      // HTTP-PUT
+      [MESSAGE_TYPES.HTTP_PUT]: (event: PostRobotEvent<any>) => {
+        return this.httpPUT(event.data.relativeURL, event.data.data, undefined, event.data.origin).then((result) => {
+          return { data: result.data, error: result.error };
+        });
+      },
+      // HTTP-DELETE
+      [MESSAGE_TYPES.HTTP_DELETE]: (event: PostRobotEvent<any>) => {
+        return this.httpDELETE(event.data.relativeURL, undefined, event.data.origin).then((result) => {
+          return { data: result.data, error: result.error };
+        });
+      },
+      // Custom Events
+      [MESSAGE_TYPES.CUSTOM_EVENT]: async (event: PostRobotEvent<any>) => {
+        if (this._eventListeners[event.data.event]) {
+          this._eventListeners[event.data.event].forEach((listener) => {
+            listener(event.data.data);
+          });
+        }
+        if (this._registeredFrames.length > 0) {
+          this._registeredFrames.forEach((frame) => {
+            // TODO: Should this make sure it doesn't echo the custom event back to the author?
```

Comment:
Would be nice to answer this question in code review before final merge. What are the pros/cons here?

Reply:
So the way I picture this is, you have three frames in a tree; A is the parent of B and C. B emits a custom event, and having no specific idea of its needs, A emits it to both B and C.

Most of the time, B probably wouldn't have code responding to this custom event. But, if B and C are the same type of frame, it's possible for them to get confused by receiving an event off of their own action. eg, if we have a "FileAddedToProfile" event, B may have already added that file itself, fired the event, and then add it a second time from seeing the event come back.

Reply:
Ahh, I see. Echoing back to the author is the easiest logic, but might not be the ideal solution. I think we can safely keep it as-is, and publishers who also want to consume the same custom events can ignore their own within their own logic.